mllib/src/main/scala/org/apache/spark/mllib/feature/IDF.scala

```diff
@@ -211,14 +213,17 @@ private object IDFModel {
     val n = v.size
     v match {
       case SparseVector(size, indices, values) =>
+        val newElements = new ArrayBuffer[(Int, Double)]
         val nnz = indices.size
-        val newValues = new Array[Double](nnz)
         var k = 0
         while (k < nnz) {
-          newValues(k) = values(k) * idf(indices(k))
+          val newValue = values(k) * idf(indices(k))
```

Comment:
If I were to follow the suggestion of updating line 58 to `new IDFModel(idf.compressed)`, then it would be possible for the `idf` vector to be sparse, but I'm not doing that here. That's part of the performance optimization I'm suggesting could happen in a separate PR.